ante/fee.go

```diff
@@ -186,85 +175,59 @@ func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
 		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "fee payer address: %s does not exist", deductFeesFrom)
 	}
 
-	var isXpla bool = false
-	for _, coin := range fee {
-		denom := coin.Denom
-
-		if denom == xplatypes.DefaultDenom {
-			isXpla = true
-		}
-	}
-
 	// deduct the fees
 	if !feeTx.GetFee().IsZero() {
+		nativeFees := sdk.Coins{}
+		xatpFees := sdk.Coins{}
 
-		if isXpla {
-			err = DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc, feeTx.GetFee())
-			if err != nil {
-				return ctx, err
-			}
-
-		} else {
+		xatpPayer := dfd.xatpKeeper.GetPayer(ctx)
 
-			ctx = ctx.WithGasMeter(sdk.NewGasMeter(sdk.Gas(dfd.smartQueryGasLimit)))
-			for _, coin := range fee {
-
-				denom := coin.Denom
-
-				ratioDec, err := dfd.xatpKeeper.GetFeeInfoFromXATP(ctx, denom)
-				if err != nil {
-					return ctx, err
-				}
-
-				var xplaDecimal sdk.Dec
-				if denom != xplatypes.DefaultDenom {
-
-					minGasPrices, err := sdk.ParseDecCoins(dfd.MinGasPrices)
-					if err != nil {
-						return ctx, err
-					}
-
-					var decimal int64
-					for _, gp := range minGasPrices {
-
-						if gp.Denom == xplatypes.DefaultDenom {
-							decimal = int64(len(gp.Amount.RoundInt().String()))
-						}
-					}
-
-					xplaDecimal = sdk.NewDecFromBigInt(new(big.Int).Mul(big.NewInt(1), new(big.Int).Exp(big.NewInt(10), big.NewInt(decimal), nil)))
+		for _, fee := range feeTx.GetFee() {
+			if xatpPayer == "" {
+				break
+			}
 
-					err = dfd.xatpKeeper.PayXATP(ctx, deductFeesFrom, denom, coin.Amount.String())
-					if err != nil {
-						return ctx, err
-					}
-				}
+			xatp, found := dfd.xatpKeeper.GetXatp(ctx, fee.Denom)
+			if !found {
+				nativeFees = nativeFees.Add(fee)
+				continue
+			}
 
-				xatpPayer := dfd.xatpKeeper.GetPayer(ctx)
-				XatpPayerAcc, err := sdk.AccAddressFromBech32(xatpPayer)
-				if err != nil {
-					return ctx, err
-				}
+			err := dfd.xatpKeeper.PayXATP(ctx, deductFeesFrom, xatp.Denom, fee.Amount.String())
+			if err != nil {
+				return ctx, err
+			}
 
-				accExists := dfd.ak.(authkeeper.AccountKeeper).HasAccount(ctx, XatpPayerAcc)
-				if !accExists {
-					return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "XATP payer account: %s not found", XatpPayerAcc)
-				}
+			ratioDec, err := dfd.xatpKeeper.GetFeeInfoFromXATP(ctx, xatp.Denom)
+			if err != nil {
+				return ctx, err
+			}
 
-				deductFeesFromAcc0 := dfd.ak.GetAccount(ctx, XatpPayerAcc)
+			feeAmount := sdk.NewDecFromIntWithPrec(fee.Amount, int64(xatp.Decimals))
+			defaultFeeAmountDec := feeAmount.Quo(ratioDec)
+			defaultFeeAmount := defaultFeeAmountDec.Mul(sdk.NewDecWithPrec(10, xplatypes.DefaultDenomPrecision)).Ceil().BigInt()
+			xatpFees = xatpFees.Add(sdk.NewCoin(xplatypes.DefaultDenom, sdk.NewIntFromBigInt(defaultFeeAmount)))
```

Comment:
The tests executes the native & CW20 transfer and I rated different gas limit
But the module withdraw same XPLA from XATP payer

Native transfer via bank module
The amount should be near 160000_000000000000 axpla (0.16 XPLA)
```
events:
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1z2k85n48ydfvzslrugwzl4j2u7vtdyf3xvucmc
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 160000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

CW20 transfer
The amount should be near 405000_000000000000 axpla (0.405 XPLA)
```
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1u27snswkjpenlscgvszcfjmz8uy2y5qacx0826
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 405000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

Reply:
```
insufficient fees; got: 160000TKN required: 680000000000000000000TKN,6800000000000000axpla: insufficient fee
```

Errored math
Let me check and derive the correct math

Reply:
https://github.com/xpladev/xpla/pull/58/commits/cf5181ea4e7dc12e916e53698ed88d410f178d18